src/platform/site-wide/user-nav/containers/Main.jsx

```diff
@@ -58,10 +54,35 @@ export class Main extends React.Component {
   }
 
   setToken = (event) => {
-    if (event.data === sessionStorage.userToken) { this.props.initializeProfile(); }
+    if (event.data === sessionStorage.userToken) {
+      this.executeRedirect();
+      this.props.initializeProfile();
+    }
   }
 
-  getRedirectUrl = () => (new URLSearchParams(window.location.search)).get('next');
+  getRedirectUrlParameter() {
+    const nextParam = (new URLSearchParams(window.location.search)).get('next');
+    if (nextParam) {
+      return nextParam.startsWith('/') ? nextParam : `/${nextParam}`;
+    }
+    return false;
+  }
+
+  getRedirectUrl = () => {
+    const nextParam = this.getRedirectUrlParameter();
+    if (nextParam) return nextParam;
+
+    return window.location.pathname === '/' && DASHBOARD_URL;
+  };
+
```

Comment:
I separated out `getRedirectUrlParameter` from `getRedirectUrl` because `getRedirectUrlParameter` is used to determine whether we should automatically open the login modal. Using `getRedirectUrl` instead would also take the user's dashboard into consideration, which means the sign-in modal would automatically open on the homepage.